Allow using DefaultRectangular arrays in GPU loops/kernels

e-kayrakli opened the pull request:
This PR allows using Chapel arrays inside GPU kernels.

Continuation of #18146 and #18199

The main challenge that this PR overcomes is passing Chapel arrays as arguments
to GPU kernels. The `_array` record is passed to functions sometimes as ref,
sometimes as val. Moreover this is determined at codegen time, so there's not
much we can do while outlining a loop body to create a GPU kernel. So, this PR
adjusts the codegen and runtime's kernel handler to take care of different
arguments differently. Mainly:

1. Codegen passes ref arguments directly (technically by value, but they are
   references), but also passes the size of the value type to the runtime. This
   allows the runtime to allocate a same sized memory on GPU, copy the
   referenced value, and pass a device pointer instead of the host reference.
2. Non-ref arguments are passed:
   a. by reference with 0 size if the argument is not of aggregate type. This
      means we're passing host references to things like integers. Runtime
      passes these references directly (because of 0 size) to `cuLaunchKernel`,
      which copies the values pointed by them into the device memory. This is
      the same semantics as passing an integer to a device kernel
   b. by reference with the size of value type if it is an aggregate type. This
      covers the case where an array in the directly outer scope is accessed in
      a `foreach`, for example. This will make the runtime behave similarly to
      case (1).

#### Other adjustments:

- Drops the number of arguments from the kernel launch primitive.
- Allows passing function symbols to the primitive. This enables having multiple
  kernels/loops to be run on the GPU.
- Drops obsolete tests, adds a new one.

I would like to rebase and merge this after #18224 and #18222.

#### Test:

- [x] `test/gpu/native`
- [x] standard

stonea commented on compiler/codegen/cg-expr.cpp:
```diff
@@ -4720,8 +4720,37 @@ static GenRet codegenGPUKernelLaunch(CallExpr* call, bool is3d) {
-      args.push_back(codegenAddrOf(actual));
+    Symbol* actualSym = toSymExpr(actual)->symbol();
+    if (curArg == 1) {
+      if (FnSymbol* fn = toFnSymbol(actualSym)) {
```
I assume this has to do with this bullet in your PR description:

> Allows passing function symbols to the primitive. This enables having multiple kernels/loops to be run on the GPU.

I guess I get line 4725: you check to see if it's a function symbol and if so you return a c string to its name. I assume this name ends up being unique-ified if we have multiple outlined kernels.

What does `isImmediate()` return?  I guess the idea behind this branch is if what's being passed in is already a C string you just codegen it and forward it?  Maybe we should assert that this argument type is something that codegen's to a CStringSymbol?


Assuming my understanding is correct could you update the comment above (starting on line 4699) to indicate the first parameter could either be the function name or a function symbol:

```C++
  // The primitive's arguments are
  //   - function name
  //   - grid size (1 arg or 3 args)
  //   - block size (1 arg or 3 args)
  //   - any number of arguments to be passed to the kernel
 ...
```

e-kayrakli replied:
Your assumptions are all correct.

> What does isImmediate() return? I guess the idea behind this branch is if what's being passed in is already a C string you just codegen it and forward it? Maybe we should assert that this argument type is something that codegen's to a CStringSymbol?

Yep, I'll add a check to make sure that the immediate is in fact a c_string.

I'll update the comment, too.

stonea commented on compiler/codegen/cg-expr.cpp:
```diff
@@ -4720,8 +4720,37 @@ static GenRet codegenGPUKernelLaunch(CallExpr* call, bool is3d) {
+        INT_FATAL("Unknown argument type in GPU launch primitive");
+      }
+    }
+    else if (curArg > nNonKernelParamArgs) {
```
Looks like for each argument we now pass the argument itself along with a size. Given that this is the case can you update the comment on line on line 4711 (tangent: I wish Github let me attach this review comment to that line), which for now says:

```C++
  // The runtime function needs the kernel parameters to be passed by address.
  // The other arguments to this primitive are passed along directly.
```

e-kayrakli commented on runtime/src/chpl-gpu.c:
```diff
@@ -181,11 +182,21 @@ static void chpl_gpu_launch_kernel_help(const char* name,
+
+    if (cur_arg_size > 0) {
+      kernel_params[i] = chpl_malloc(1*sizeof(CUdeviceptr));
+      *kernel_params[i] = chpl_gpu_mem_alloc(cur_arg_size, 0, 0, 0);
```
For the description, I want to wait for a little while to see what type of descriptors we need. For context, we use those descriptors to be more descriptive in verbose memory setups and things like memory leaks. Note that we also have a `chpl_malloc` and `chpl_free` in this same file that needs to use `chpl_mem_alloc` and `chpl_mem_free` instead along with proper descriptors that is probably different than this. But they have `TODO`s and this should have, too :)

I have a toy branch where I pass the line numbers to the `chpl_gpu_launch_kernel_flat` from the compiler. I'll get that merged soon, and then we can add line/file info, then.

e-kayrakli commented:
@stonea  -- Thanks! I believe I've addressed all your comments. If you can take a look at the last batch of commits that I added (probably one by one), that'd be great. Meanwhile, I am firing a `release/examples`, non-gpu paratest. `test/gpu/native` passes after the changes.
